<html>
<body>
<h3>Data</h3>
<br/>
<br/>
<div class="well">
  <h4>Individual-level breadth of coverage per CCDS transcript for 1000 Genomes exome data</h4>
  <ul>
    <li><a href="downloads/coverage_b37.tar.gz" target="_blank">Coverage data in b37</a></li>
    <li><a href="downloads/coverage_hg38.tar.gz" target="_blank">Coverage data in hg38</a></li>
  </ul>
  <p>Each consists of separate <i><a href="https://github.com/fstpackage/fst" target="_blank">fst</a></i> files for target depths: 5x, 10x, 15x, 20x, 25x, 30x, 50x, 100x.</p>
</div>
<div class="well">
  <h4>Exome coverage summary from gnomAD (release 2.1, b37 only)</h4>
  <ul>
    <li><a href="downloads/gnomad_exome_cov.tar.gz" target="_blank">gnomAD exome coverage summary</a></li>
  </ul>
  <p>Fraction of gnomAD exomes over 5x, 10x, 15x, 20x, 25x, 30x, 50x, 100x for each locus in CCDS genes (in <i><a href="https://genome.ucsc.edu/goldenPath/help/bigWig.html" target="_blank">bigwig</a></i> format).</p>
</div>
<div class="well">
  <h4>Other data necessary to run <i>WEScover</i></h4>
  <ul>
    <li><a href="downloads/data.tar.gz" target="_blank">Data for app</a></li>
  </ul>
  <p>Contains all other data necessary to run <i>WEScover</i> locally.</p>
</div>
<div class="well">
  <h4>Source</h4>
  <ul>
    <li>The source code of this Shiny app can be found at <a href="https://github.com/bch-gnome/WEScover" target="_blank">GitHub</a></li>
  </ul>
</div>

</body>
</html>
